### ISSUE TITLE
Is there a difference between dgl.ops.edge_softmax and torch.sparse.softmax?

### ISSUE THREAD

**Issue by maqy1995** [dmlc/dgl#2311]
## ❓ Questions and Help

dgl.ops.edge_softmax link : https://docs.dgl.ai/generated/dgl.ops.edge_softmax.html?highlight=edge_softmax#dgl.ops.edge_softmax

torch.sparse.softmax link:https://pytorch.org/docs/master/sparse.html?highlight=torch%20sparse%20softmax#torch.sparse.softmax

In my understanding, the effect of `dgl.ops.edge_softmax` and `torch.sparse.softmax` should be the same, so I run the code below, but the result is different. What is wrong?
```
# -*- coding: utf-8 -*-
import dgl
import torch
from dgl.data.knowledge_graph import load_data
from dgl.ops import edge_softmax

"""
@Author: maqy1995
@Time: 2020/10/30
@Description: test `dgl.ops.edge_softmax` and `torch.sparse.softmax`
"""

# load graph data
data = load_data('FB15k-237')
num_nodes = data.num_nodes
train_data = data.train
valid_data = data.valid
test_data = data.test
num_rels = data.num_rels

# build graph
g = dgl.graph([])
g.add_nodes(num_nodes)
src, rel, dst = train_data.transpose()
g.add_edges(src, dst)

print(g)

# create relation features
relation_embed = torch.nn.Embedding(num_rels, 1)
torch.nn.init.xavier_normal_(relation_embed.weight)

edge_features = relation_embed(torch.from_numpy(rel)).squeeze()

# traditional sparse matrix
train_data = torch.LongTensor(train_data)
a_matrix = torch.sparse.FloatTensor(train_data[:, [0, 2]].T, edge_features, torch.Size([num_nodes, num_nodes]))

a_matrix_torch = torch.sparse.softmax(a_matrix, dim=0)

# dgl edge_softmax
attention_from_edge_softmax = edge_softmax(g, edge_features)

a_matrix_dgl = torch.sparse.FloatTensor(train_data[:, [0, 2]].T, attention_from_edge_softmax, torch.Size([num_nodes, num_nodes]))

res_torch = a_matrix_torch.detach().to_dense()
res_dgl = a_matrix_dgl.detach().to_dense()
print("res_torch equal to res_dgl:" + str(res_torch.equal(res_dgl)))

print("res_torch[0][0] = {}".format(res_torch[0][0]))
print("res_dgl[0][0] = {}".format(res_dgl[0][0]))

print("res_torch[0][1042] = {}".format(res_torch[0][1042]))
print("res_dgl[0][1042] = {}".format(res_dgl[0][1042]))

print("-" * 50)

print("res_torch[3364][8619] = {}".format(res_torch[3364][8619]))
print("res_dgl[3364][8619] = {}".format(res_dgl[3364][8619]))

print("res_torch[14540][13094] = {}".format(res_torch[14540][13094]))
print("res_dgl[14540][13094] = {}".format(res_dgl[14540][13094]))

print("res_torch[14540][13665] = {}".format(res_torch[14540][13665]))
print("res_dgl[14540][13665] = {}".format(res_dgl[14540][13665]))

```
the output is:
```
Graph(num_nodes=14541, num_edges=272115,
      ndata_schemes={}
      edata_schemes={})
res_torch equal res_dgl:False
res_torch[0][0] = 0.20204603672027588
res_dgl[0][0] = 0.16857339441776276
res_torch[0][1042] = 0.12746837735176086
res_dgl[0][1042] = 0.11301705986261368
--------------------------------------------------
res_torch[3364][8619] = 0.014925372786819935
res_dgl[3364][8619] = 0.014925372786819935
res_torch[14540][13094] = 0.0025641026441007853
res_dgl[14540][13094] = 0.0025641026441007853
res_torch[14540][13665] = 0.0007508183480240405
res_dgl[14540][13665] = 0.0007181554101407528

Process finished with exit code 0
```
Looking forward to your reply.

**Comment by yzh119** [comment]
Hi, we did not notice pytorch has supported its own sparse softmax before, seems it's new feature of PyTorch 1.7.

If you are dealing with scalar shape edge feature, they are equivalent:
```python
>>> import torch
import>>> import dgl
Using backend: pytorch
RDFLib Version: 5.0.0
>>> g = dgl.graph(([0, 0, 0, 1, 1, 2], [0, 1, 2, 1, 2, 2]))
>>> i = torch.stack(g.edges(), 0)
>>> edata = torch.rand(6)
>>> sp = torch.sparse.FloatTensor(i, edata, torch.Size([3, 3]))
>>> sp
tensor(indices=tensor([[0, 0, 0, 1, 1, 2],
                       [0, 1, 2, 1, 2, 2]]),
       values=tensor([0.2819, 0.1179, 0.1877, 0.0746, 0.5700, 0.7742]),
       size=(3, 3), nnz=6, layout=torch.sparse_coo)
>>> torch.sparse.softmax(sp, 0)
tensor(indices=tensor([[0, 0, 0, 1, 1, 2],
                       [0, 1, 2, 1, 2, 2]]),
       values=tensor([1.0000, 0.5108, 0.2345, 0.4892, 0.3438, 0.4217]),
       size=(3, 3), nnz=6, layout=torch.sparse_coo)
>>> torch.sparse.softmax(sp, 1)
tensor(indices=tensor([[0, 0, 0, 1, 1, 2],
                       [0, 1, 2, 1, 2, 2]]),
       values=tensor([0.3625, 0.3077, 0.3299, 0.3786, 0.6214, 1.0000]),
       size=(3, 3), nnz=6, layout=torch.sparse_coo)
>>> dgl.ops.edge_softmax(g, edata)
tensor([1.0000, 0.5108, 0.2345, 0.4892, 0.3438, 0.4217])
>>> dgl.ops.edge_softmax(g, edata, norm_by='src')
tensor([0.3625, 0.3077, 0.3299, 0.3786, 0.6214, 1.0000])
```

softmax according to dim 0 is equivalent to dgl's edge_softmax normalized by destination node, and softmax according to dim 1 is equivalent to dgl's edge_softmax normalized by source node.

**Comment by yzh119** [comment]
The advantage of DGL's edge softmax is that PyTorch's sparse matrix is stored in COO format while DGL maintain variable sparse formats and select the one with best performance for different operators. 

For edge softmax, using CSC format is more efficient and we can guarantee deterministic property, while PyTorch's current implementation is based on COO format: https://github.com/pytorch/pytorch/blob/29dc3c5ec821f5b9026e1c847c0ac605672e95af/aten/src/ATen/native/sparse/SoftMax.cpp

**Comment by maqy1995** [comment]
thank you for your detailed reply.
What puzzles me is that theoretically, the results of these two methods should be the same, but as I posted, some of the outputs are the same, but some are different:
same output:
```
res_torch[3364][8619] = 0.014925372786819935
res_dgl[3364][8619] = 0.014925372786819935
res_torch[14540][13094] = 0.0025641026441007853
res_dgl[14540][13094] = 0.0025641026441007853
```
different output:
```
res_torch[0][0] = 0.20204603672027588
res_dgl[0][0] = 0.16857339441776276
res_torch[0][1042] = 0.12746837735176086
res_dgl[0][1042] = 0.11301705986261368
```

I am not sure if it is my code problem, the environment I used is dgl-cu102 0.5.2, and pytorch 1.6/1.7(I try both).

**Comment by yzh119** [comment]
Is the graph you experimented on a multiple graph (i.e. there are duplicate edges)?

**Comment by maqy1995** [comment]
I am not sure，as you can see，I construct graph from FB15k-237 dataset with dgl’s api.




| |
马卿云
邮箱：maqy1995@163.com
|

签名由 网易邮箱大师 定制

On 10/30/2020 20:00, Zihao Ye wrote:

Is the graph you experimented on a multiple graph (i.e. there are duplicate edges)?

**Comment by yzh119** [comment]
@maqy1995 I manually checked the edge softmax value on edges connected to node 0 and I think DGL's answer is right.
```python
>>> eids = []
>>> val = []
>>> us = []
>>> for i, (u, v) in enumerate(zip(*g.edges())):
...     if v == 0:
...             eids.append(i)
...             val.append(edge_features[i].item())
... 
>>> us
[tensor(2548), tensor(1042), tensor(4994), tensor(6786), tensor(0), tensor(6786)]
>>> eids
[79109, 79301, 82498, 119779, 183397, 194623]
>>> vals
[-0.11333584785461426, -0.11333584785461426, -0.11333584785461426, 0.14713697135448456, 0.09009554237127304, 0.04487883299589157]
>>> torch.softmax(torch.tensor(vals), 0)
tensor([0.1494, 0.1494, 0.1494, 0.1938, 0.1831, 0.1750])
>>> attention_from_edge_softmax[eids]
tensor([0.1494, 0.1494, 0.1494, 0.1938, 0.1831, 0.1750],
       grad_fn=<IndexBackward>)
>>> a_matrix_torch[0, 0]
tensor(0.2195, grad_fn=<SelectBackward>)  # wrong answer
```

**Comment by maqy1995** [comment]
thanks a lot! Maybe PyTorch needs a new issue to confirm this.

**Comment by maqy1995** [comment]
today I checked it again, and I found you are right, there are duplicate edges in FB15k-237, which influence the result.
```
index_to_node0_from_dgl_edges = []
vals_dgl_edge = []
for i, (u, v) in enumerate(zip(*g.edges())):
  ...:     if v == 0:
  ...:         vals_dgl_edge.append(edge_features[i].item())
  ...:         index_to_node0_from_dgl_edges.append(u)
  ...:         

index_to_node0_from_dgl_edges
Out[11]: 
[tensor(2548),
 tensor(1042),
 tensor(4994),
 tensor(6786),
 tensor(0),
 tensor(6786)]

vals_dgl_edge
Out[10]: 
[0.09308641403913498,
 0.09308641403913498,
 0.09308641403913498,
 0.07495848834514618,
 -0.12471813708543777,
 0.055945463478565216]
 
index_to_node0_from_sparse_matrix = []
vals_sparse_matrix = []
for i, x in enumerate(matrix_dense[:, 0]):
   ...:      if x != 0:
   ...:          vals_sparse_matrix.append(x.detach().item())
   ...:          index_to_node0_from_sparse_matrix.append(i)
   ...:          
index_to_node0_from_sparse_matrix:
Out[17]: [0, 1042, 2548, 4994, 6786] # there only one value from node 6786 to 
node 0 when we use sparse matrix, but there are two edges from node 6786 to node 0.

vals_sparse_matrix:
Out[18]: 
[-0.12471813708543777,
 0.09308641403913498,
 0.09308641403913498,
 0.09308641403913498,
 0.130903959274292]

```
we can found that there only one value from node 6786 to node 0 when we use sparse matrix(which value is `0.130903959274292 = 0.07495848834514618 + 0.055945463478565216`), but there are two edges from node 6786 to node 0.

In the case of multiple graph, I think the calculation result of DGL is more reasonable.
Thanks again for your detailed reply.
